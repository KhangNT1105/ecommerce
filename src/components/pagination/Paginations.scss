.paginations-wrapper {
  display: flex;
  justify-content: space-between;

  .page-link {
    outline: none;
    box-shadow: none;
  }

  .paginations-container {
    display: flex;
    justify-content: flex-end;

    .go-to-container {
      display: flex;
      padding-right: 20px;

      p {
        font-size: 15px;
        padding-right: 10px;
        padding-top: 3px;
      }

      input {
        height: 31px;
        width: 31px;
        padding-left: 10px;
      }
    }
  }

  .show-results-container {
    display: flex;
  }
}

@media (max-width: 576px){
  .paginations-wrapper {
    display: block;
  }
}
