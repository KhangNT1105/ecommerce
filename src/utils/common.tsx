import React from 'react';
import { Denomination } from 'Props.d';
import { assign, clone, map, reduce, isObjectLike } from 'lodash';
import { validateRequired, validateEmail } from 'utils/validation';
import { ACTIVATE, ENTITY_STATUS_FROM_SERVER, DEACTIVATE, DELETE, APPROVE, REJECT } from '../constants';
import { Genre, SubGenre } from 'stores/ConfigurationStore/configurationType.d';

export const checkRenderDoubleValue = (value1?: string | number, value2?: string | number, value3?: any | '') => {
  let minus = '';
  if (value1 && value2) {
    minus = ' - ';
  }
  return (
    <>
      {value1} {value3 && <b>{value3}</b>}
      {minus}
      {value2}
    </>
  );
};

export type SelectOptionProps = {
  id: string;
  value: string;
  label: string;
};

type DataProps = {
  [key: string]: any;
};

export type SelectOptionKeyMappingProps = {
  key: string;
  value: string;
  label: string;
};

export const renderSelectOptions = (data: DataProps[], keyMapping?: SelectOptionKeyMappingProps) => {
  return (
    <>
      <option value="">Select</option>
      {data.map((item) => {
        if (keyMapping && Object.entries(keyMapping).length > 0) {
          return (
            <option key={item[keyMapping.key]} value={item[keyMapping.value]}>
              {item[keyMapping.label]}
            </option>
          );
        }
        return (
          <option key={item.id} value={item.value}>
            {item.label}
          </option>
        );
      })}
    </>
  );
};

export const checkRenderArray = (items: string[]) => {
  return <span>{items.join(', ')}</span>;
};

export const getListPrices = (denominations: Denomination[]) => {
  const listPrices = denominations.map((item) => item.price);
  return listPrices.sort((a, b) => a - b);
};
export const getPrices = (productDetail: any) => {
  if (productDetail.valueType === 'RANGE') {
    return checkRenderDoubleValue(productDetail.minPrice, productDetail.maxPrice);
  }
  if (productDetail.valueType === 'SLAB') {
    const values = getListPrices(productDetail.denominations);
    return values.join(', ');
  }
};

export interface IPagingModelOption {
  pageIndex: number;
  pageSize: number;
}

export interface IPagingModelResult<T> {
  items: T[];
  hasPrevPage: boolean;
  hasNextPage: boolean;
  prevPage: number | null;
  nextPage: number | null;
  size: number;
  page: number;
  totalItems: number;
  totalPages: number;
}

export function getPaginationResponse<T>(
  pagingModel: IPagingModelOption,
  result: T[],
  totalItems: number
): IPagingModelResult<T> {
  const totalPages = Math.floor((totalItems + pagingModel.pageSize - 1) / pagingModel.pageSize);
  const pagingModelResult: IPagingModelResult<T> = {
    items: result,
    totalItems,
    totalPages,
    hasPrevPage: pagingModel.pageIndex > 0,
    hasNextPage: pagingModel.pageIndex < totalPages - 1,
    prevPage: pagingModel.pageIndex > 0 ? pagingModel.pageIndex - 1 : null,
    nextPage: pagingModel.pageIndex < totalPages - 1 ? pagingModel.pageIndex + 1 : null,
    size: pagingModel.pageSize,
    page: pagingModel.pageIndex
  };
  return pagingModelResult;
}

export function replaceSpecialCharacters(str: string, character: string) {
  return str.replace(/(?:\.(?![^.]+$)|[^\w.])+/g, character);
}

export const classConcat = (classes: string[]): string => {
  if (classes.length) {
    return classes.filter((className) => !!className).join(' ');
  }
  return '';
};

export const recursiveGetFieldName = (obj: object, entityField: string, fieldNameCombine?: string): any => {
  let listFieldConfig: any;
  listFieldConfig = reduce(
    obj,
    (result, value, key) => {
      let cloneResult = clone(result);
      let localFieldName: any = clone(fieldNameCombine);
      if (key === entityField) {
        map(value, (item: any, k) => {
          const getFieldName = `${localFieldName}.${key}.${k}`;

          // check whether any validation
          if (item.validationRules && item.validationRules.length > 0) {
            item.validationRules.map((itemValidate: any, index: number) => {
              if (itemValidate.rule === 'isRequired') {
                item.validationRules[index].validator = validateRequired;
              }
              if (itemValidate.rule === 'isEmail') {
                item.validationRules[index].validator = validateEmail;
              }

              return itemValidate;
            });
          }

          // insert key, value into list field
          return (cloneResult = {
            ...cloneResult,
            [getFieldName]: item
          });
        }); // reset field name

        localFieldName = '';
      } else if (isObjectLike(value)) {
        if (!localFieldName) {
          localFieldName = key;
        } else {
          localFieldName += `.${key}`;
        } // recall recursive fn if not be found the expected entityField

        return assign(cloneResult, recursiveGetFieldName(value, entityField, localFieldName));
      }

      listFieldConfig = {
        ...listFieldConfig,
        ...cloneResult
      };

      return listFieldConfig;
    },
    {}
  );

  return listFieldConfig;
};

export type ObjectFieldType = {
  [field: string]: string | object | string[] | object[] | any;
};

export const trimAllValueWithObjectField = (objField: ObjectFieldType) => {
  if ((!Array.isArray(objField) && typeof objField !== 'object') || objField === null) {
    return objField;
  }
  return Object.keys(objField).reduce(
    (acc: ObjectFieldType, key) => {
      acc[key] =
        typeof objField[key] === 'string'
          ? (objField[key] as string).trim()
          : trimAllValueWithObjectField(objField[key]);
      return acc;
    },
    Array.isArray(objField) ? [] : {}
  );
};

const listShowBtnActivate = [
  ENTITY_STATUS_FROM_SERVER.INACTIVATE,
  ENTITY_STATUS_FROM_SERVER.DRAFT,
  ENTITY_STATUS_FROM_SERVER.UPDATED,
  ENTITY_STATUS_FROM_SERVER.ACTIVATE
];

const listShowBtnSMActivate = [
  ENTITY_STATUS_FROM_SERVER.INACTIVATE,
  ENTITY_STATUS_FROM_SERVER.DRAFT,
  ENTITY_STATUS_FROM_SERVER.UPDATED
];

const listShowBtnSMDelete = [ENTITY_STATUS_FROM_SERVER.DRAFT, ENTITY_STATUS_FROM_SERVER.INACTIVATE];

const listShowBtnDeactivate = [ENTITY_STATUS_FROM_SERVER.ACTIVATE, ENTITY_STATUS_FROM_SERVER.UPDATED];

const listShowBtnDelete = [ENTITY_STATUS_FROM_SERVER.INACTIVATE, ENTITY_STATUS_FROM_SERVER.DRAFT];

export const checkEntityActionNonApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnActivate.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivate.includes(entityStatus);
    case DELETE:
      return listShowBtnDelete.includes(entityStatus);
    default:
      return true;
  }
};

export const checkSMEntityActionNonApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnSMActivate.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivate.includes(entityStatus);
    case DELETE:
      return listShowBtnSMDelete.includes(entityStatus);
    default:
      return true;
  }
};

export const getAllSubGenres = (genres: Genre[]) =>
  genres
    .map(({ children }) => children)
    .reduce((currentValue, subGenres) => {
      if (subGenres) {
        return [...currentValue, ...subGenres];
      }
      return currentValue;
    }, []);

export const getAllMicroGenres = (subGenres: SubGenre[]) =>
  subGenres
    .map(({ children }) => children)
    .reduce((currentValue, microGenre) => {
      if (microGenre) {
        return [...currentValue, ...microGenre];
      }
      return currentValue;
    }, []);

const listShowBtnActivateWithApproval = [...listShowBtnActivate, ENTITY_STATUS_FROM_SERVER.REJECT];
const listShowBtnDeactivateWithApproval = [...listShowBtnDeactivate];
const listShowBtnDeleteWithApproval = [...listShowBtnDelete];
const listShowBtnApproveWithApproval = [ENTITY_STATUS_FROM_SERVER.APPROVAL_PENDING];
const listShowBtnRejectWithApproval = [ENTITY_STATUS_FROM_SERVER.APPROVAL_PENDING];

export const checkEntityActionApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnActivateWithApproval.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivateWithApproval.includes(entityStatus);
    case DELETE:
      return listShowBtnDeleteWithApproval.includes(entityStatus);
    case APPROVE:
      return listShowBtnApproveWithApproval.includes(entityStatus);
    case REJECT:
      return listShowBtnRejectWithApproval.includes(entityStatus);
    default:
      return true;
  }
};

export function createMarkup() {
  return {
    __html: `<div class="zalo-chat-widget" data-oaid="927003371386666481" data-welcome-message="Rất vui khi được hỗ trợ bạn!" data-autopopup="0" data-width="450" data-height="450"></div>
    `
  };
}
