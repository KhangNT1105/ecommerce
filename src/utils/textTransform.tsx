export const upperCaseFirst = (str: String = '') => {
  return str.toLowerCase().replace(/^\w/, (chars) => chars.toUpperCase());
};
