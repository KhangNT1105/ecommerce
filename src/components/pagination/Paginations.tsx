import React, { MouseEvent, useEffect, useState } from 'react';

import { Pagination, PaginationItem, PaginationLink } from 'reactstrap';
import { useTranslation } from 'react-i18next';
import { paginationsProps } from './Paginations.d';
import { LIST_MAX_ROWS, PAGINATION } from '../../constants/index';
import ShowRow from './showRow/ShowRow';
import ShowResults from './showResults/ShowResults';
import './Paginations.scss';
import REGEX from 'constants/regex';

const Paginations: React.FC<paginationsProps> = ({ handlePagination, isShowGoTo, totals, isUpdatedData }) => {
  const pages: number[] = [];
  const { t } = useTranslation();
  const defaultPageSize = PAGINATION.DEFAULT_PAGE_SIZE;
  const defaultCurrentPage = PAGINATION.DEFAULT_CURRENT_PAGE;
  const [valueInput, setValue]: any = useState(1);
  const [isChangeGoto, setChangeGoto] = useState(false);
  const [pager, setPager] = useState({
    currentPage: 1,
    pageSize: 15,
    totalItems: 0,
    startPage: 0,
    endPage: 0,
    startIndex: 0,
    totalPages: 0,
    endIndex: 0,
    from: 1,
    to: 15,
    pages
  });
  const [currentSize, setCurrentSize] = useState(15);

  const getPager = (totalItems: number, currentPage: number, pageSize: number) => {
    const newCurrentPage = currentPage || defaultCurrentPage;

    // default page size is 25
    const newPageSize = pageSize || defaultPageSize;
    const totalPages = Math.ceil(totalItems / pageSize);
    let startPage: number = 1;
    let endPage: number = 1;
    if (totalPages <= PAGINATION.NUMBER_OF_NAVIGATION_SHOW) {
      // less than 10 total pages so show all
      startPage = 1;
      endPage = totalPages;
    } else {
      // more than 10 total pages so calculate start and end pages
      if (currentPage <= PAGINATION.BEGIN_TO_PAGING) {
        startPage = 1;
        endPage = PAGINATION.NUMBER_OF_NAVIGATION_SHOW;
      } else if (currentPage + PAGINATION.UPPER_BOUND >= totalPages) {
        startPage = totalPages - PAGINATION.NUMBER_OF_NAVIGATION_SHOW + 1;
        endPage = totalPages;
      } else {
        startPage = currentPage - PAGINATION.LOWER_BOUND;
        endPage = currentPage + PAGINATION.UPPER_BOUND;
      }
    }

    const startIndex = (currentPage - 1) * pageSize;
    const endIndex = Math.min(startIndex + pageSize - 1, totalItems - 1);

    let from: number = 1;
    let to: number = pageSize;

    if (currentPage > 1 && currentPage < totalPages) {
      from = pageSize * (currentPage - 1) + 1;
      to = pageSize * currentPage;
    }

    if (currentPage === totalPages) {
      from = pageSize * (currentPage - 1) + 1;
      to = totalItems;
    }
    // create an array of pages to ng-repeat in the pager control
    const pages = [...Array(endPage + 1 - startPage).keys()].map((i) => startPage + i);

    // return object with all pager properties required by the view
    return {
      totalItems,
      currentPage: newCurrentPage,
      pageSize: newPageSize,
      totalPages,
      startPage,
      endPage,
      startIndex,
      endIndex,
      pages,
      from,
      to
    };
  };

  const setPage = (page: number) => {
    // get new pager object for specified page
    const newPager = getPager(totals, page, currentSize);

    // update state
    setPager(newPager);
    handlePagination && handlePagination(page, currentSize);
  };

  useEffect(() => {
    if (totals) {
      const newPager = getPager(totals, 1, currentSize);
      setPager(newPager);
      setChangeGoto(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [totals, currentSize]);

  useEffect(() => {
    if (isUpdatedData) {
      setPage(1);
      setValue(1);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isUpdatedData]);

  const handlePageClick = (e: MouseEvent, page: number) => {
    e.preventDefault();
    setPage(page);
    setValue(page);
    handlePagination && handlePagination(page, currentSize);
  };

  const handlePreviousClick = (e: MouseEvent) => {
    e.preventDefault();
    setPage(pager.currentPage - 1);
    setValue(pager.currentPage - 1);
    handlePagination && handlePagination(pager.currentPage - 1, currentSize);
  };

  const handleNextClick = (e: MouseEvent) => {
    e.preventDefault();
    setPage(pager.currentPage + 1);
    setValue(pager.currentPage + 1);
    handlePagination && handlePagination(pager.currentPage + 1, currentSize);
  };

  const handleFirstPageClick = (e: MouseEvent) => {
    e.preventDefault();
    setPage(1);
    setValue(1);
    handlePagination && handlePagination(0, currentSize);
  };

  const handleLastPageClick = (e: MouseEvent) => {
    e.preventDefault();
    setPage(pager.totalPages);
    setValue(pager.totalPages);
    handlePagination && handlePagination(pager.totalPages, currentSize);
  };

  const handleBlur = (e: React.FormEvent<HTMLInputElement>) => {
    const value = e.currentTarget.value;

    if (!REGEX.DIGIT_ONLY.test(value)) {
      setValue(pager.currentPage);
      return;
    }

    if (Number(value) > pager.totalPages) {
      setPage(pager.totalPages);
      setValue(pager.totalPages);
      handlePagination && handlePagination(pager.totalPages, currentSize);
    }
    if (Number(value) <= 1) {
      setPage(1);
      setValue(1);
      handlePagination && handlePagination(0, currentSize);
    }

    if (Number(value) >= 1 && Number(value) <= pager.totalPages) {
      setValue(value);
      setPage(Number(value));
      handlePagination && handlePagination(Number(value), currentSize);
    }
  };

  const handleToGoChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const pageNumber: number = parseInt(event.target.value, 10);
    setChangeGoto(true);
    pageNumber > pager.totalPages ? setValue(pager.totalPages) : setValue(pageNumber);
  };

  const handleFocus = () => {
    setValue('');
  };
  const paginationItems = pager.pages.map((page: number, i) => (
    <PaginationItem active={page === pager.currentPage} key={i}>
      <PaginationLink onClick={(e) => handlePageClick(e, page)} href="#">
        {page}
      </PaginationLink>
    </PaginationItem>
  ));

  return (
    <div className="paginations-wrapper">
      <div className="show-results-container">
        <ShowRow
          listMaxRows={LIST_MAX_ROWS}
          currentSize={currentSize}
          setCurrentSize={setCurrentSize}
          handlePagination={handlePagination}
        />
        <ShowResults page={pager.currentPage} size={pager.totalPages} />
      </div>
      <div className="paginations-container">
        {isShowGoTo && (
          <div className="go-to-container">
            <p>
              <span>{t('GO_TO')}</span>
            </p>
            <input
              type="number"
              onBlur={handleBlur}
              onFocus={handleFocus}
              onChange={handleToGoChange}
              value={isChangeGoto ? valueInput : pager.currentPage}
            />
          </div>
        )}
        <Pagination aria-label="Page navigation example" size="5">
          <PaginationItem disabled={pager.currentPage <= 1} className="first-pagination">
            <PaginationLink first={true} onClick={handleFirstPageClick} href="#">
              <i className="fa fa-fw fa-angle-double-left" />
            </PaginationLink>
          </PaginationItem>
          <PaginationItem disabled={pager.currentPage <= 1} className="prev-pagination">
            <PaginationLink previous={true} onClick={handlePreviousClick} href="#">
              <i className="fa fa-fw fa-angle-left" />
            </PaginationLink>
          </PaginationItem>
          {paginationItems}
          <PaginationItem disabled={pager.currentPage >= pager.totalPages} className="next-pagination">
            <PaginationLink next={true} onClick={handleNextClick} href="#">
              <i className="fa fa-fw fa-angle-right" />
            </PaginationLink>
          </PaginationItem>
          <PaginationItem disabled={pager.currentPage >= pager.totalPages} className="last-pagination">
            <PaginationLink last={true} onClick={handleLastPageClick} href="#">
              <i className="fa fa-fw fa-angle-double-right" />
            </PaginationLink>
          </PaginationItem>
        </Pagination>
      </div>
    </div>
  );
};

Paginations.defaultProps = {
  isShowGoTo: true
};

export default Paginations;
