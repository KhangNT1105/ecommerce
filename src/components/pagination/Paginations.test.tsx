import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { testSnapshots } from '../../utils/test';
import Paginations from './Paginations';

const exampleItems = [...Array(300).keys()].map((i) => ({
  id: i + 1,
  name: `Item + ${i + 1}`
}));

const defaultProps = {
  handlePagination: jest.fn(),
  totals: exampleItems.length
};

const e = {
  preventDefault: jest.fn()
};
describe('<Paginations />', () => {
  testSnapshots(Paginations, [
    {
      props: { ...defaultProps },
      description: 'render snapshot correctly'
    },
    {
      props: { ...defaultProps, isShowGoTo: false },
      description: 'do not render go To when isShowGoTo is false'
    }
  ]);
});

describe('Test Logic', () => {
  afterEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
    jest.resetModules();
  });

  test('should set to go to maximum page if current value is larger than maximum page', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const instance = mount(<Paginations {...props} />);
    instance.find('.page-link').at(3).simulate('click');
    instance.mount();
    instance.setProps({ totals: 5 });
    expect((instance.find('.go-to-container').children('input').instance() as any).value).toBe('1');
  });

  test('should set to go to page 1 if page size is set', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const event = {
      target: {
        value: '3'
      }
    };
    jest.doMock('./showRow/ShowRow', () => 'ShowRow');
    const instance = mount(<Paginations {...props} />);
    instance.find('input').simulate('change', event);
    (instance.find('ShowRow').props() as any).setCurrentSize(50);
    instance.mount();

    expect(instance.find('input').props().value).toBe(1);
  });

  test('should go to previous page when click previous pagination', () => {
    const setState = jest.fn();
    const useStateMock: any = (initState: any) => [initState, setState];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const props = {
      ...defaultProps
    };
    const wrapper = shallow(<Paginations {...props} />);
    wrapper.find('.prev-pagination').children('PaginationLink').simulate('click', e);
    expect(setState).toHaveBeenCalledWith({
      currentPage: 1,
      endIndex: -1,
      endPage: 5,
      from: 1,
      pageSize: 15,
      pages: [1, 2, 3, 4, 5],
      startIndex: -15,
      startPage: 1,
      to: 15,
      totalItems: 300,
      totalPages: 20
    });
    expect(setState).toHaveBeenCalledWith(0);
  });
  test('should go to first page when click first pagination', () => {
    const setState = jest.fn();
    const useStateMock: any = (initState: any) => [initState, setState];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const props = {
      ...defaultProps
    };
    const wrapper = shallow(<Paginations {...props} />);
    wrapper.find('.first-pagination').children('PaginationLink').simulate('click', e);
    expect(setState).toHaveBeenCalledWith({
      currentPage: 1,
      endIndex: 14,
      endPage: 5,
      from: 1,
      pageSize: 15,
      pages: [1, 2, 3, 4, 5],
      startIndex: 0,
      startPage: 1,
      to: 15,
      totalItems: 300,
      totalPages: 20
    });
    expect(setState).toHaveBeenCalledWith(1);
  });

  test('should go to next page when click next pagination', () => {
    const setState = jest.fn();
    const useStateMock: any = (initState: any) => [initState, setState];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const props = {
      ...defaultProps
    };
    const wrapper = shallow(<Paginations {...props} />);
    wrapper.find('.next-pagination').children('PaginationLink').simulate('click', e);
    expect(setState).toHaveBeenCalledWith({
      currentPage: 2,
      endIndex: 29,
      endPage: 5,
      from: 16,
      pageSize: 15,
      pages: [1, 2, 3, 4, 5],
      startIndex: 15,
      startPage: 1,
      to: 30,
      totalItems: 300,
      totalPages: 20
    });
    expect(setState).toHaveBeenCalledWith(2);
  });

  test('should go to last page when click last pagination', () => {
    const setState = jest.fn();
    const useStateMock: any = (initState: any) => [initState, setState];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const props = {
      ...defaultProps
    };
    const wrapper = shallow(<Paginations {...props} />);
    wrapper.find('.last-pagination').children('PaginationLink').simulate('click', e);
    expect(setState).toHaveBeenCalledWith({
      currentPage: 1,
      endIndex: -1,
      endPage: 5,
      from: 1,
      pageSize: 15,
      pages: [1, 2, 3, 4, 5],
      startIndex: -15,
      startPage: 1,
      to: 15,
      totalItems: 300,
      totalPages: 20
    });
    expect(setState).toHaveBeenCalledWith(0);
  });

  test('should set value undefine when focus go to ', () => {
    const setState = jest.fn();
    const useStateMock: any = (initState: any) => [initState, setState];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const wrapper = shallow(<Paginations {...props} />);
    wrapper.find('.go-to-container').children('input').simulate('focus');
    expect(setState).toHaveBeenCalledWith('');
  });

  test('should set value undefine when focus go to ', () => {
    const setState = jest.fn();
    const useStateMock: any = (initState: any) => [initState, setState];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const wrapper = shallow(<Paginations {...props} />);
    wrapper.find('.go-to-container').children('input').simulate('focus');
    expect(setState).toHaveBeenCalledWith('');
  });

  describe('go to box handling', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };

    const instance = mount(<Paginations {...props} />);
    instance.update();
    instance.mount();
    const input = instance.find('.go-to-container').children('input');

    const onBlurInput = (input.props() as any).onBlur;
    afterEach(() => {
      defaultProps.handlePagination.mockClear();
    });

    test('should call maximum page when the value exeeds it', () => {
      onBlurInput({
        currentTarget: {
          value: '1000'
        }
      });
      expect(defaultProps.handlePagination).toHaveBeenCalledWith(20, 15);
    });

    test('should not call handlePagination when input negative number', () => {
      onBlurInput({
        currentTarget: {
          value: '-1'
        }
      });
      expect(defaultProps.handlePagination).not.toHaveBeenCalled();
    });

    test('should call value when it is between 1 and maximum page exclusively', () => {
      onBlurInput({
        currentTarget: {
          value: '3'
        }
      });
      expect(defaultProps.handlePagination).toHaveBeenCalledWith(3, 15);
    });
  });

  test('should return the page list from 2 -> 6 when page 4 is chosen', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const instance = mount(<Paginations {...props} />);
    const setPage = (instance.find('.page-item .page-link').at(5).props() as any).onClick;
    setPage({ preventDefault: jest.fn() });
    instance.mount();

    expect(instance.find('.page-item .page-link').at(2).text()).toBe('2');
    expect(instance.find('.page-item .page-link').at(6).text()).toBe('6');
  });

  test('should return the page list from 1 -> 5 when page 3 is chosen', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const instance = mount(<Paginations {...props} />);
    const setPage = (instance.find('.page-item .page-link').at(4).props() as any).onClick;
    setPage({ preventDefault: jest.fn() });
    instance.mount();

    expect(instance.find('.page-item .page-link').at(2).text()).toBe('1');
    expect(instance.find('.page-item .page-link').at(6).text()).toBe('5');
  });

  test('should return the page list from 6->10 when page 8 is chosen', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const instance = mount(<Paginations {...props} />);
    (instance.find('input').props() as any).onBlur({
      currentTarget: { value: '8' }
    });
    instance.mount();

    expect(instance.find('.page-item .page-link').at(2).text()).toBe('6');
    expect(instance.find('.page-item .page-link').at(6).text()).toBe('10');
  });

  test('should return the page list from 8 -> 20(last page) when page 11 is chosen', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const instance = mount(<Paginations {...props} />);
    (instance.find('input').props() as any).onBlur({
      currentTarget: { value: '11' }
    });
    instance.mount();

    expect(instance.find('.page-item .page-link').at(2).text()).toBe('9');
    expect(instance.find('.page-item .page-link').at(6).text()).toBe('13');
  });

  test('should not change current page list when input invalid value', () => {
    const props = {
      ...defaultProps,
      isShowGoTo: true
    };
    const instance = mount(<Paginations {...props} />);
    defaultProps.handlePagination.mockClear();

    (instance.find('input').props() as any).onBlur({
      currentTarget: { value: 'abcde' }
    });
    instance.mount();
    expect(defaultProps.handlePagination).not.toHaveBeenCalled();

    expect(instance.find('.page-item .page-link').at(2).text()).toBe('1');
    expect(instance.find('.page-item .page-link').at(6).text()).toBe('5');
  });
});
