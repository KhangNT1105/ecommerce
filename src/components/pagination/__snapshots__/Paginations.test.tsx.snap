// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginations /> snapshots do not render go To when isShowGoTo is false 1`] = `
<div
  className="paginations-wrapper"
>
  <div
    className="show-results-container"
  >
    <div
      className="show-row-container"
    >
      <p>
        <span>
          SHOW_ROWS
        </span>
      </p>
      <div
        className="dropup dropdown"
        onKeyDown={[Function]}
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          aria-label={null}
          className="dropdown-toggle btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          15
        </button>
        <div
          aria-hidden={true}
          className="dropdown-menu"
          role="menu"
          tabIndex="-1"
        >
          <button
            className="dropdown-item"
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
            type="button"
          >
            15
          </button>
          <button
            className="dropdown-item"
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
            type="button"
          >
            18
          </button>
          <button
            className="dropdown-item"
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
            type="button"
          >
            21
          </button>
        </div>
      </div>
    </div>
    <div
      className="show-results"
    >
      SHOW_RESULTS
    </div>
  </div>
  <div
    className="paginations-container"
  >
    <nav
      aria-label="Page navigation example"
      className=""
    >
      <ul
        className="pagination pagination-5"
      >
        <li
          className="first-pagination page-item disabled"
        >
          <a
            aria-label="First"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-double-left"
              />
            </span>
            <span
              className="sr-only"
            >
              First
            </span>
          </a>
        </li>
        <li
          className="prev-pagination page-item disabled"
        >
          <a
            aria-label="Previous"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-left"
              />
            </span>
            <span
              className="sr-only"
            >
              Previous
            </span>
          </a>
        </li>
        <li
          className="page-item active"
        >
          <a
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            1
          </a>
        </li>
        <li
          className="page-item"
        >
          <a
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            2
          </a>
        </li>
        <li
          className="page-item"
        >
          <a
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            3
          </a>
        </li>
        <li
          className="page-item"
        >
          <a
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            4
          </a>
        </li>
        <li
          className="page-item"
        >
          <a
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            5
          </a>
        </li>
        <li
          className="next-pagination page-item"
        >
          <a
            aria-label="Next"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-right"
              />
            </span>
            <span
              className="sr-only"
            >
              Next
            </span>
          </a>
        </li>
        <li
          className="last-pagination page-item"
        >
          <a
            aria-label="Last"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-double-right"
              />
            </span>
            <span
              className="sr-only"
            >
              Last
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`<Paginations /> snapshots render snapshot correctly 1`] = `
<div
  className="paginations-wrapper"
>
  <div
    className="show-results-container"
  >
    <div
      className="show-row-container"
    >
      <p>
        <span>
          SHOW_ROWS
        </span>
      </p>
      <div
        className="dropup dropdown"
        onKeyDown={[Function]}
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          aria-label={null}
          className="dropdown-toggle btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          15
        </button>
        <div
          aria-hidden={true}
          className="dropdown-menu"
          role="menu"
          tabIndex="-1"
        >
          <button
            className="dropdown-item"
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
            type="button"
          >
            15
          </button>
          <button
            className="dropdown-item"
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
            type="button"
          >
            18
          </button>
          <button
            className="dropdown-item"
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
            type="button"
          >
            21
          </button>
        </div>
      </div>
    </div>
    <div
      className="show-results"
    >
      SHOW_RESULTS
    </div>
  </div>
  <div
    className="paginations-container"
  >
    <div
      className="go-to-container"
    >
      <p>
        <span>
          GO_TO
        </span>
      </p>
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="number"
        value={1}
      />
    </div>
    <nav
      aria-label="Page navigation example"
      className=""
    >
      <ul
        className="pagination pagination-5"
      >
        <li
          className="first-pagination page-item disabled"
        >
          <a
            aria-label="First"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-double-left"
              />
            </span>
            <span
              className="sr-only"
            >
              First
            </span>
          </a>
        </li>
        <li
          className="prev-pagination page-item disabled"
        >
          <a
            aria-label="Previous"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-left"
              />
            </span>
            <span
              className="sr-only"
            >
              Previous
            </span>
          </a>
        </li>
        <li
          className="next-pagination page-item disabled"
        >
          <a
            aria-label="Next"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-right"
              />
            </span>
            <span
              className="sr-only"
            >
              Next
            </span>
          </a>
        </li>
        <li
          className="last-pagination page-item disabled"
        >
          <a
            aria-label="Last"
            className="page-link"
            href="#"
            onClick={[Function]}
          >
            <span
              aria-hidden="true"
            >
              <i
                className="fa fa-fw fa-angle-double-right"
              />
            </span>
            <span
              className="sr-only"
            >
              Last
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
